File: app/controllers/Authentication.java
package controllers;

import com.google.common.base.Joiner;
import com.google.common.collect.Lists;
import models.User;
import org.apache.commons.lang3.StringUtils;
import play.Logger;
import play.data.Form;
import play.mvc.Controller;
import play.mvc.Result;
import play.mvc.Results;
import views.html.login;

import java.util.List;

import static play.data.Form.form;

/**
 * @author f.patin
 */
public class Authentication extends Controller {

	public static class LoginForm {
		public String username;
		public String password;
		public String redirect;
	}

	public static Result login() {
		return ok(login.render());
	}

	public static Result authenticate() {
		final Form<LoginForm> loginForm = form(LoginForm.class).bindFromRequest();
		if (loginForm.hasErrors()) {
			return badRequest();
		}
		final String username = loginForm.get().username;
		final String password = loginForm.get().password;
		final String redirect = loginForm.get().redirect;


		if (User.findAuthorisedUser(username, password) != null) {
			session("username", username);
			final List<String> url = Lists.newArrayList(routes.Application.index().url());
			if (StringUtils.isNotBlank(redirect)) {
				url.add("#" + redirect);
			}
			return Results.redirect(Joiner.on("").join(url));
		} else {
			Logger.error(String.format("%s.authenticate failure user:[%s]", Authentication.class.getSimpleName(), username));
			session().clear();
			flash("errormsg", "Utilisateur ou mot de passe incorrect");
			return Results.redirect(routes.Authentication.login());
		}
	}

	public static Result logout() {
		session().clear();
		return Results.redirect(routes.Authentication.login());
	}
}

File: app/controllers/Authentication.scala
package controllers

import play.api.Logger
import play.api.data.Form
import play.api.data.Forms.{text, optional, tuple, nonEmptyText}
import play.api.libs.functional.syntax._
import play.api.libs.json._
import play.api.mvc.{Action, Security, Results}
import security.{Profile, MyDeadboltHandler, Auth}
import utils.MD5
import http.CacheControl
import reactivemongo.bson.BSONObjectID
import play.modules.reactivemongo.json.BSONFormats._
import play.api.libs.json.JsString
import scala.Some

/**
 * @author f.patin
 */
object Authentication extends BaseController {

  def login = Action {
    implicit request =>
      Ok(views.html.login())
  }

  def logout = Action {
    request =>
      Results.Redirect(routes.Authentication.login()).withSession(request.session - Security.username)
  }

  val authenticateForm = Form(tuple(
    "username" -> nonEmptyText,
    "password" -> nonEmptyText,
    "redirect" -> optional(text)
  ))

  def authenticate = Action {
    implicit request =>
      authenticateForm.bindFromRequest.fold(
        withErrors => {
          Results.Redirect(routes.Authentication.login())
            .withSession(request.session - Security.username)
            .flashing(("errormsg" -> "Utilisateur ou mot de passe incorrect"))
        },
        form => {
          Async {
            Auth.checkAuthentication((form._1, MD5(form._2))).map {
              check =>
                check match {
                  case Some(auth) =>
                    form._3.map(next => Results.Redirect(routes.Application.index() + "#" + next))
                      .getOrElse(Results.Redirect(routes.Application.index()))
                      .withSession(request.session + (Security.username -> auth.username))
                  case None => Results.Redirect(routes.Authentication.login())
                    .withSession(request.session - Security.username)
                    .flashing(("errormsg" -> "Utilisateur ou mot de passe incorrect"))
                }
            }.recover {
              case e =>
                Logger.error("authenticate", e)
                InternalServerError(JsString(s"exception ${e.getMessage}"))
            }
          }
        }
      )

  }

  val fromProfile: OWrites[Profile] = (
    (__ \ "id").writeNullable[String] and
      (__ \ "username").write[String] and
      (__ \ "role").write[String]
    )(unlift(Profile.unapply))

  def profile = Restrict(everybody, new MyDeadboltHandler()) {
    Action {
      request =>
        Async {
          Profile(request.session.get("username").get)
            .map {
            profile => Ok(Json.toJson(profile.get)(fromProfile)).withHeaders(profile.get.eTag, CacheControl.maxAgeO)
          }
            .recover {
            case e: Exception => InternalServerError(JsString(s"exception ${e.getMessage}"))
          }
        }
    }

  }
}
